src/components/swap/SwapHeader.tsx

```diff
@@ -13,14 +22,120 @@ const StyledSwapHeader = styled.div`
+`
+
+const TextHeader = styled.div<{ color: string; marginLeft?: string; isClickable?: boolean; disabled?: boolean }>`
+  color: ${({ color }) => color};
+  opacity: ${({ disabled }) => disabled && '60%'};
+  margin-left: ${({ marginLeft }) => marginLeft};
+  margin-right: 8px;
+  display: flex;
+  line-height: 20px;
+  flex-direction: row;
+  cursor: ${({ isClickable }) => isClickable && 'pointer'};
+  justify-content: center;
+  align-items: center;
+`
+
+const MOONPAY_REGION_AVAILABILITY_ARTICLE =
+  'https://support.uniswap.org/hc/en-us/articles/11306664890381-Why-isn-t-MoonPay-available-in-my-region-'
+
+// TODO(lynnshaoyu): add analytics and logging
 export default function SwapHeader({ allowedSlippage }: { allowedSlippage: Percent }) {
+  const theme = useTheme()
+  const { account } = useWeb3React()
+  const openFiatOnRampModal = useOpenModal(ApplicationModal.FIAT_ONRAMP)
+  const fiatOnRampButtonEnabled = useFiatOnRampButtonEnabled()
+  const [buyFiatClicked, setBuyFiatClicked] = useBuyFiatClicked()
+  const [continueBuyFiatFlowPostRegionCheck, setContinueBuyFiatFlowPostRegionCheck] = useState(false)
+  const [continueBuyFiatFlowPostWalletModal, setContinueBuyFiatFlowPostWalletModal] = useState(false)
+  const [checkFiatRegionAvailability, setCheckFiatRegionAvailability] = useState(false)
+  const {
+    available: fiatOnrampAvailable,
+    availabilityChecked: fiatOnrampAvailabilityChecked,
+    error,
+    loading: fiatOnrampAvailabilityLoading,
+  } = useFiatOnrampAvailability(checkFiatRegionAvailability)
+  const [walletDrawerOpen, toggleWalletDrawer] = useWalletDrawer()
+
+  const handleBuyCrypto = useCallback(() => {
+    if (!fiatOnrampAvailabilityChecked) {
+      setCheckFiatRegionAvailability(true)
+      setContinueBuyFiatFlowPostRegionCheck(true)
+    } else if (fiatOnrampAvailable && !account && !walletDrawerOpen) {
+      toggleWalletDrawer()
+      setContinueBuyFiatFlowPostRegionCheck(false)
+      setContinueBuyFiatFlowPostWalletModal(true)
+    } else if (fiatOnrampAvailable && account) {
+      openFiatOnRampModal()
+      setBuyFiatClicked(true)
+      setContinueBuyFiatFlowPostRegionCheck(false)
+      setContinueBuyFiatFlowPostWalletModal(false)
+    } else if (!fiatOnrampAvailable) {
+      setBuyFiatClicked(true)
+      setContinueBuyFiatFlowPostRegionCheck(false)
+      setContinueBuyFiatFlowPostWalletModal(false)
+    }
+  }, [
+    walletDrawerOpen,
+    fiatOnrampAvailabilityChecked,
+    fiatOnrampAvailable,
+    account,
+    toggleWalletDrawer,
+    openFiatOnRampModal,
+    setBuyFiatClicked,
+  ])
+
+  useEffect(() => {
+    if (continueBuyFiatFlowPostRegionCheck || (account && continueBuyFiatFlowPostWalletModal)) {
+      handleBuyCrypto()
+    }
+  }, [account, handleBuyCrypto, continueBuyFiatFlowPostWalletModal, continueBuyFiatFlowPostRegionCheck])
+
+  const disableBuyCryptoButton = Boolean(
+    error || (!fiatOnrampAvailable && fiatOnrampAvailabilityChecked) || fiatOnrampAvailabilityLoading
```

Comment:
also does this `error` handling work correctly if it's a one-off error like internet issues? it seems like from this code that there's no way to 'reset' the error by clicking on the button again

Reply:
good point, changed it to log error to console